### ISSUE TITLE
bisecting modules (was Re: [perl #XXX] Segfault on syntax checking)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11732]
Migrated from [rt.perl.org#102948](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=102948) (status was 'rejected')

Searchable as RT102948$

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Nov 04\, 2011 at 08&#8203;:27&#8203;:16AM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Nov 04 06&#8203;:13&#8203;:26 2011\, nicholas wrote&#8203;:
> > I wrote the bisect tool in the hope that other people would use it too
> > :\-/
> 
> I used it\! Once\, that is\, resulting in
> \<https://rt.cpan.org/Ticket/Display.html?id=72158>.

I know Merijn did too\.

> But I find it awfully hard to figure out how to install modules while
> running it and how to find the commit that caused a warning to start
> appearing\.  If you could make it more like Andreas König's
> binsearchaperl\, it would be even easier to use\.

Historically I've always tended to try to first reduce any problem with
the module to the smallest test case I can \(hopefully\, by then\, without
needing to use the module or any of its dependencies\)\. Partly\, I guess
because I'm going to have to do that anyway\, if I want a test case to add
to the core\, and partly because it's easier to examine things under gdb \(or
equivalent\) when there's less Perl code being processed\. Which means that
I've tended to assume that I'd have a small enough test case to feed to
git bisect\.

I'm not sure of a good way to automate installation of 1\+ modules as
part of a bisect run\. I think it's going to have to be

\* install the perl you just built
\* run its CPAN shell to install modules

except that that fels fragile at the best of times\, and will go wrong on
earlier versions of the CPAN shell that didn't default to automatic
configuration\. So maybe it's "flatten the dependency tree to a list of modules
to install" before starting\, then build that list in the run\.

Either way\, I'm not completely sure that I can automate it well enough to
work\, and I was scared by how much time it turned out I've already spent on
the bisect tool\. It was spread over 6 weekly reports\, so I hadn't summed
them until generating the October monthly report\. I knew why I did what I
did each day or two\, but I hadn't realised the bigger picture\. So I don't
think that it's the right thing for me to work on in the near future\.

Nicholas Clark

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'rejected'

**Comment by p5pRT** [comment]
### From @nwc10

Yes\, I didn't mean to create a new bug\. Whoops\.

**Comment by p5pRT** [comment]
### From @ikegami

On Fri\, Nov 4\, 2011 at 3&#8203;:38 PM\, Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:

> \# New Ticket Created by  Nicholas Clark
> \# Please include the string&#8203;:  \[perl \#102948\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=102948 >
>
>
> On Fri\, Nov 04\, 2011 at 08&#8203;:27&#8203;:16AM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> > But I find it awfully hard to figure out how to install modules while
> > running it and how to find the commit that caused a warning to start
> > appearing\.  If you could make it more like Andreas König's
> > binsearchaperl\, it would be even easier to use\.
>
> Historically I've always tended to try to first reduce any problem with
> the module to the smallest test case I can
>

I tried to use the tool and was unable to figure out how to have it install
a module\. I was trying to use the bisect tool to identify the commit that
broke my XS module\. Reduction wasn't really possible since 1\) the module
was rather trivial\, and 2\) I would still be left with an XS module\.

I couldn't even build the XS module without installing \(not just making\)
Perl because Perl was reporting that its file were somewhere they weren't
yet\.

Thankfully\, I had nearly identical functional code and was able to identify
the problem through comparison of the two&#8203;: Creating a list op with no
children isn't allowed\.

\- Eric

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Nov 04 16&#8203;:15&#8203;:15 2011\, ikegami@&#8203;adaelis\.com wrote&#8203;:
> Thankfully\, I had nearly identical functional code and was able to
> identify
> the problem through comparison of the two&#8203;: Creating a list op with no
> children isn't allowed\.

That’s strange\, because perl does that itself\.  See 958ed56b\, for instance\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Fri Nov 04 16&#8203;:15&#8203;:15 2011\, ikegami@&#8203;adaelis\.com wrote&#8203;:
> Thankfully\, I had nearly identical functional code and was able to
> identify
> the problem through comparison of the two&#8203;: Creating a list op with no
> children isn't allowed\.

That’s strange\, because perl does that itself\.  See 958ed56b\, for instance\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @craigberry

On Fri\, Nov 4\, 2011 at 6&#8203;:14 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:
> On Fri\, Nov 4\, 2011 at 3&#8203;:38 PM\, Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>
> wrote&#8203;:
>>
>> \# New Ticket Created by  Nicholas Clark
>> \# Please include the string&#8203;:  \[perl \#102948\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=102948 >
>>
>>
>> On Fri\, Nov 04\, 2011 at 08&#8203;:27&#8203;:16AM \-0700\, Father Chrysostomos via RT
>> wrote&#8203;:
>> > But I find it awfully hard to figure out how to install modules while
>> > running it and how to find the commit that caused a warning to start
>> > appearing\.  If you could make it more like Andreas König's
>> > binsearchaperl\, it would be even easier to use\.
>>
>> Historically I've always tended to try to first reduce any problem with
>> the module to the smallest test case I can
>
> I tried to use the tool and was unable to figure out how to have it install
> a module\. I was trying to use the bisect tool to identify the commit that
> broke my XS module\. Reduction wasn't really possible since 1\) the module was
> rather trivial\, and 2\) I would still be left with an XS module\.
>
> I couldn't even build the XS module without installing \(not just making\)
> Perl because Perl was reporting that its file were somewhere they weren't
> yet\.


I'm probably missing something obvious\, but ExtUtils&#8203;::ParseXS runs
tests that build and run mini XS extensions during the core build\.  Is
it really necessary to install an extension to test it rather than
just writing a script that does what those tests do?  I\.e\.\, starting
from here&#8203;:

\<http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/dist/ExtUtils\-ParseXS/t/001\-basic\.t>

making a copy of the utterly trivial XSTest\.xs\, replacing the guts of
it with the guts of the "rather trivial" extension in question\, and
doctoring the test script to exercise whatever misbehavior has been
observed?

In other words\, if bisect\.pl can bisect a Perl script and if a Perl
script can create\, build\, and run an XS extension\, why can't those two
things be put together?

You might need relatively recent ExtUtils&#8203;::ParseXS and
ExtUtils&#8203;::CBuilder to do this\, which would limit how far back you
could go\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon\, Nov 14\, 2011 at 05&#8203;:14&#8203;:05PM \-0600\, Craig A\. Berry wrote&#8203;:
> I'm probably missing something obvious\, but ExtUtils&#8203;::ParseXS runs
> tests that build and run mini XS extensions during the core build\.  Is
> it really necessary to install an extension to test it rather than
> just writing a script that does what those tests do?  I\.e\.\, starting
> from here&#8203;:
> 
> \<http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/dist/ExtUtils\-ParseXS/t/001\-basic\.t>
> 
> making a copy of the utterly trivial XSTest\.xs\, replacing the guts of
> it with the guts of the "rather trivial" extension in question\, and
> doctoring the test script to exercise whatever misbehavior has been
> observed?
> 
> In other words\, if bisect\.pl can bisect a Perl script and if a Perl
> script can create\, build\, and run an XS extension\, why can't those two
> things be put together?
> 
> You might need relatively recent ExtUtils&#8203;::ParseXS and
> ExtUtils&#8203;::CBuilder to do this\, which would limit how far back you
> could go\.

The problem is you need to install that module's entire dependency
tree the same way\, and since this is in an automated bisect\, it all
needs to be automated\.

Since it all just works with an installed perl we might as well
install it\.

Tony

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Nov 15\, 2011 at 10&#8203;:26&#8203;:39AM \+1100\, Tony Cook wrote&#8203;:

> The problem is you need to install that module's entire dependency
> tree the same way\, and since this is in an automated bisect\, it all
> needs to be automated\.
> 
> Since it all just works with an installed perl we might as well
> install it\.

Yes\, this was partly my thinking\.

Also\, historically\, not all extensions have built properly when unpacked
into ext/
We'd be getting false negatives in this case\, which would be confusing
\(ie reported failures\, which should not be failures\)

Nicholas Clark
